---
layout: post
title: python urllib内存泄露问题分析
category: python
catalog: true
published: true
tags:
  - python
time: '2023.02.23 12:07:00'
---

# 一、问题背景
在python2/3中调用urllib模块会有循环引用没被释放，触发代码如下所示：
```python
import gc
import sys


if sys.version_info.major == 3:
    from urllib import request
    urlopen = request.urlopen
else:
    import urllib2
    urlopen = urllib2.urlopen


gc.collect()
# the unreachable object will be freed if this flag not set
gc.set_debug(gc.DEBUG_SAVEALL)
print("before call urlopen(), the unreachable object len: %s" % len(gc.garbage))

a = urlopen('http://www.google.com')
del a
# check memory on memory leaks
gc.collect()
print("after call urlopen(), the unreachable object len: %s" % len(gc.garbage))
```
在python2.7执行示例代码可以看到如下输出：
```shell
before call urlopen(), the unreachable object len: 0
after call urlopen(), the unreachable object len: 21
```
在python3.9执行示例代码可以看到如下输出：
```shell
before call urlopen(), the unreachable object len: 0
after call urlopen(), the unreachable object len: 0
```
从触发用例输出结果初步看到python2.7的urllib2模块有内存泄露问题。

# 二、urllib2模块代码分析
修改测试代码对引用泄露进行分析，修改后的测试代码如下所示：
```python
import gc
import sys

import objgraph


if sys.version_info.major == 3:
    from urllib import request
    urlopen = request.urlopen
else:
    import urllib2
    urlopen = urllib2.urlopen


gc.collect()
# Set the gc can print the leak object.
gc.set_debug(gc.DEBUG_LEAK)
print("Before calling urlopen(), the unreachable object len: %s" % len(gc.garbage))

a = urlopen('http://www.google.com')
del a
# check memory on memory leaks
print("After calling urlopen(), collecting the unreachable objects: %s" % gc.collect())
```
输出结果如下所示，从这里看可以发现有21个不可达对象被回收掉。
```shell
Before calling urlopen(), the unreachable object len: 0
gc: collectable <HTTPResponse instance at 0x7f91f8614050>
gc: collectable <dict 0x7f91f85be3b0>
gc: collectable <HTTPMessage instance at 0x7f91f8cabbe0>
gc: collectable <dict 0x7f91f85be170>
gc: collectable <list 0x7f91f8600280>
gc: collectable <list 0x7f91f8602690>
gc: collectable <instancemethod 0x7f91f8600e10>
gc: collectable <HTTPResponse instance at 0x7f91f85b3190>
gc: collectable <dict 0x7f91f85be950>
gc: collectable <HTTPMessage instance at 0x7f91f85b30f0>
gc: collectable <dict 0x7f91f85be710>
gc: collectable <list 0x7f91f85b3280>
gc: collectable <list 0x7f91f85b31e0>
gc: collectable <instancemethod 0x7f91f8602b90>
gc: collectable <HTTPResponse instance at 0x7f91f85b3c80>
gc: collectable <dict 0x7f91f85c5170>
gc: collectable <HTTPMessage instance at 0x7f91f85b3d20>
gc: collectable <dict 0x7f91f85beef0>
gc: collectable <list 0x7f91f85b3b90>
gc: collectable <list 0x7f91f85b3c30>
gc: collectable <instancemethod 0x7f91f85b3730>
After calling urlopen(), collecting the unreachable objects: 21
```

# 三、参考文献
- [urllib2's urlopen() method causes a memory leak](https://github.com/python/cpython/issues/42012)
- [https://www.cnblogs.com/anjike/p/10230302.html](https://www.zhihu.com/question/21747929/answer/569469557)
- [urlopen内存泄漏浅析](https://www.cnblogs.com/anjike/p/10230302.html)
- [objgraph: Memory leak example](https://mg.pov.lt/objgraph/#memory-leak-example)